### PR TITLE
Resolve #2707: Lucene directory: treat close as a potential recovery path

### LINKED ISSUE #2707: Lucene directory: treat close as a potential recovery path

### PR DESCRIPTION
**Pull request by jjezra**

### CODE REVIEW

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryWrapper.java [hunk 1]
```diff
@@ -291,19 +294,30 @@ public IndexWriter getWriter(@Nonnull LuceneAnalyzerWrapper analyzerWrapper, @Nu
...
+            }
+        }
+    }
+
     @Override
     @SuppressWarnings("PMD.CloseResource")
     public synchronized void close() throws IOException {
         if (writer != null) {
-            writer.close();
+            tryClose(writer, "writer");
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryWrapper.java:312]: This should still throw an exception if any of them fail to close, but it should proceed to the next one. You could use `IOUtils.close(writer, writerReader, directory)` which handles this for you.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java [hunk 2]
```diff
@@ -452,8 +452,8 @@ public <R> CompletableFuture<R> apply(Function<FDBRecordContext, CompletableFutu
 
         @Override
         public <R> CompletableFuture<R> applyInRecoveryPath(Function<FDBRecordContext, CompletableFuture<R>> function) {
-            // No recovery for a single user transaction
-            return CompletableFuture.completedFuture(null);
+            // Best effort - skip ensureOpen, ignore exceptions.
+            return function.apply(callerContext).exceptionally(ex -> null);
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java:456]: I'm assuming this change was found necessary by the additional coverage in the test?

**Review by ScottDugas — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java (hunk 2 above)
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java:456]: Yes. Without it the non-agile version had failed.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryWrapper.java (hunk 1 above)
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryWrapper.java:312]: Nice! 
✅

**Review by ScottDugas — APPROVED** [review]

**Pull request merged**